### DISCUSSION TITLE
Best practices to add commit messages

### DISCUSSION THREAD

**Discussion by CoderMaster62** [recodehive/Stackoverflow-Analysis#459] [Q&A] [answered]
Best practices to add commit messages

**Comment by SameerNadaf** [comment, marked as answer]
Writing clear and meaningful commit messages is essential for maintaining a clean and understandable project history. Here are some best practices for creating effective commit messages:

---

### 1. **Follow a Consistent Format**
Use a standard format for your commit messages. The most common style is:
```
<type>: <short summary>

<optional detailed description>
```
For example:
```
feat: add user authentication module

Implemented login, signup, and password reset functionality using JWT. Updated tests to cover new endpoints.
```

#### Common Commit Types:
- **feat**: A new feature.
- **fix**: A bug fix.
- **docs**: Documentation updates (e.g., README changes).
- **style**: Code style changes (no functionality changes, e.g., formatting).
- **refactor**: Code refactoring without adding features or fixing bugs.
- **test**: Adding or updating tests.
- **chore**: Maintenance tasks (e.g., updating dependencies).

---

### 2. **Keep Messages Short and Concise**
- **Title**: Limit the title to **50 characters** or less. This is the first line of the commit message and should be a brief summary.
- **Description**: If necessary, add a more detailed explanation after a blank line, limited to **72 characters per line**.

---

### 3. **Use the Imperative Mood**
Write the commit message as if you're giving a command or describing what this commit will do. Examples:
- ✅ "Add login feature" (imperative mood)
- ❌ "Added login feature" (past tense)

This aligns with how Git describes changes (e.g., *"Apply patch"*).

---

### 4. **Explain the Why and What**
- **What**: What changes were made?
- **Why**: Why were these changes necessary?
- Avoid explaining *how* the changes were implemented unless it's particularly complex.

For example:
```
fix: prevent crash on null user input

Fixed a bug where the app would crash when a null value was passed
to the user input handler. Added a null check to ensure stability.
```

---

### 5. **Group Related Changes**
Each commit should focus on a single task or change. Avoid combining unrelated updates into one commit.

---

### 6. **Avoid WIP Commits**
Avoid committing work-in-progress (WIP) changes. Use `git stash` to temporarily save changes instead. If necessary, use a draft message like:
```
WIP: Add basic structure for dashboard
```
...but plan to squash or clean it up before merging.

---

### 7. **Reference Issues and PRs**
If the commit relates to an issue or pull request, reference it:
```
fix: resolve login redirect bug (#123)
```

Or:
```
feat: implement search functionality

Closes #456
```

---

### 8. **Use Tools to Standardize Commit Messages**
- Use tools like **Commitizen** to enforce a consistent commit message structure.
- Set up a **Git hook** with tools like `commitlint` to validate commit messages.

---

### 9. **Avoid Noise in Messages**
Skip vague messages like:
- "Fix stuff"
- "Update code"
- "Change file"

Instead, specify what was fixed or updated.

---

### 10. **Review Your Message**
Before committing, read the message as if you’re another developer. Does it make sense? Would it help you understand the change?

---

### Example of a Good Commit Message:
```
feat: add password reset functionality

Added a password reset feature to the user authentication module.
Users can now request a password reset link via email. Updated the
email templates and added tests for the password reset flow.
```

By following these best practices, you’ll improve the clarity and quality of your project’s history, making collaboration smoother for everyone involved!
